sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -1605,4 +1623,293 @@ public String generateSas(BlobServiceSasSignatureValues blobServiceSasSignatureV
+        return queryWithResponse(expression, null)
+            .flatMapMany(BlobQueryAsyncResponse::getValue);
+    }
+
+    /**
+     * Queries the entire blob.
+     *
+     * <p>For more information, see the
+     * <a href="https://docs.microsoft.com/en-us/rest/api/">Azure Docs</a></p>
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.blob.specialized.BlobAsyncClientBase.queryWithResponse#String-BlobQueryOptions}
+     *
+     * @param expression The query expression.
+     * @param queryOptions {@link BlobQueryOptions The query options}.
+     * @return A reactive response containing the queried data.
+     */
+    public Mono<BlobQueryAsyncResponse> queryWithResponse(String expression, BlobQueryOptions queryOptions) {
+        try {
+            return withContext(context ->
+                queryWithResponse(expression, queryOptions, context));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    Mono<BlobQueryAsyncResponse> queryWithResponse(String expression, BlobQueryOptions queryOptions, Context context) {
+
+        BlobQueryOptions finalQueryOptions = queryOptions == null ? new BlobQueryOptions() : queryOptions;
+        BlobRequestConditions requestConditions = finalQueryOptions.getRequestConditions() == null
+            ? new BlobRequestConditions() : finalQueryOptions.getRequestConditions();
+
+        QuickQuerySerialization in = transformSerialization(finalQueryOptions.getInputSerialization(), logger);
+        QuickQuerySerialization out = transformSerialization(finalQueryOptions.getOutputSerialization(), logger);
+
+        QueryRequest qr = new QueryRequest()
+            .setExpression(expression)
+            .setInputSerialization(in)
+            .setOutputSerialization(out);
+
+        return this.azureBlobStorage.blobs().quickQueryWithRestResponseAsync(null, null, qr,
+            getSnapshotId(), null, requestConditions.getLeaseId(), requestConditions.getIfModifiedSince(),
+            requestConditions.getIfUnmodifiedSince(), requestConditions.getIfMatch(),
+            requestConditions.getIfNoneMatch(), null, getCustomerProvidedKey(), context)
+            .map(response -> new BlobQueryAsyncResponse(response.getRequest(), response.getStatusCode(),
+                response.getHeaders(),
+                parse(response.getValue(), o -> this.parseRecord(o, finalQueryOptions.getErrorReceiver(),
+                    finalQueryOptions.getProgressReceiver())),
+                response.getDeserializedHeaders()));
+    }
+
+    /**
+     * Avro parses a quick query reactive stream.
+     *
+     * @param avro The reactive stream.
+     * @param objectHandler The function to handle objects.
+     * @return The parsed quick query reactive stream.
+     */
+    private Flux<ByteBuffer> parse(Flux<ByteBuffer> avro, Function<Object, Mono<ByteBuffer>> objectHandler) {
```

Comment:
Can you add a description of what's happening here and what the data looks like? Like we get a stream of bytes that's avro. It's expected to have this schema. Once we get the data out of it, we need to apply these transformations, etc.

Reply:
Also, if this is the main BlobClient, I think it might be best to put the Query private helper methods somewhere else, so we don't clutter it.

Reply:
Yeah I can move these into a BlobQueryUtil in implementation